File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        
        # 2 2 2
        # 0 0 2
        # 2 2 0
        maximum, minimum = nums[:], nums[:]
        for i in range(1, len(nums)):
            maximum[i] = max(maximum[i - 1], maximum[i])
        for i in range(len(nums) - 2, -1, -1):
            minimum[i] = min(minimum[i], minimum[i + 1])
        index = float("inf")
        for i in range(len(nums)):
            if maximum[i] - minimum[i] <= k:
                index = min(index, i)
        return index if index != float("inf") else -1

File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var firstStableIndex = function(nums, k) {
    var maxLeft = Array.from({length : nums.length}, (_, i) => nums[i])
    var minRight = Array.from({length : nums.length}, (_, i) => nums[i])
    for(let i = 1; i < maxLeft.length; i++) {
        maxLeft[i] = Math.max(maxLeft[i], maxLeft[i - 1])
    }
    for(let i = minRight.length - 2; i > -1; i--) {
        minRight[i] = Math.min(minRight[i], minRight[i + 1])
    }
    for(let i = 0; i < nums.length; i++) {
        if(maxLeft[i] - minRight[i] <= k) {
            return i
        }
    }
    return -1
};